File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        //for explanation check the cpp soln
        int maxlen = -1;
        int n = s.length();
        int start = 0;
        for(int i=0;i<n;i++) {
            int evenpal = expandAroundCenter(s,i,i+1);
            int oddpal = expandAroundCenter(s,i,i);
            int len = Math.max(evenpal,oddpal);
            if(len>maxlen) {
                maxlen = len;
                start = i - (maxlen - 1) / 2;
            }
        }
        return s.substring(start,start+maxlen);
    }
    public int expandAroundCenter(String s, int left, int right) {
        while(left>=0 && right<s.length() && s.charAt(left)==s.charAt(right)) {
            left--;
            right++;
        }
        return right-left-1;
    }
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        if (s.empty()) return "";

        int start = 0, maxLength = 1; // Variables to store the starting index and maximum length of palindrome found.

        for (int i = 0; i < s.length(); i++) {
            // Check for odd-length palindromes (single character center)
            int len1 = expandAroundCenter(s, i, i);
            // Check for even-length palindromes (two character center)
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = max(len1, len2);

            // Update maximum length and starting index if a longer palindrome is found
            if (len > maxLength) {
                maxLength = len;
                start = i - (maxLength - 1) / 2; // Calculate the new starting index
            }
        }

        return s.substr(start, maxLength); // Return the longest palindromic substring
    }

private:
    int expandAroundCenter(const string &s, int left, int right) {
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--; // Expand left
            right++; // Expand right
        }
        return right - left - 1; // Return the length of the palindrome
    }
};